thirdparty/seq/sequence.hpp

```diff
@@ -0,0 +1,420 @@
+//***************************************************************************
+//* Copyright (c) 2015 Saint Petersburg State University
+//* Copyright (c) 2011-2014 Saint Petersburg Academic University
+//* All Rights Reserved
+//* See file LICENSE for details.
+//***************************************************************************
+
+#pragma once
+
+#include "log.hpp"
+#include "nucl.hpp"
+#include "utils/sfinae_checks.hpp"
+
+#include <llvm/ADT/IntrusiveRefCntPtr.h>
+#include <llvm/ADT/SparseBitVector.h>
+#include <llvm/Support/TrailingObjects.h>
+
+#include <vector>
+#include <string>
+#include <memory>
+#include <sstream>
+#include <cstring>
+
+// Silence bogus gcc warnings
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wconversion"
+
+class Sequence {
+    // Type to store Seq in Sequences
+    typedef uint64_t ST;
+    // Number of bits in ST
+    const static size_t STBits = sizeof(ST) << 3;
+    // Number of nucleotides in ST
+    const static size_t STN = (STBits >> 1);
+    // Number of bits in STN (for faster div and mod)
+    const static size_t STNBits = log_<STN, 2>::value;
+
+    class ManagedNuclBuffer final : public llvm::ThreadSafeRefCountedBase<ManagedNuclBuffer>,
+                                    protected llvm::TrailingObjects<ManagedNuclBuffer, ST> {
+        friend TrailingObjects;
+
+        ManagedNuclBuffer() {}
+
+        ManagedNuclBuffer(size_t nucls, ST *buf) {
+            std::uninitialized_copy(buf, buf + Sequence::DataSize(nucls), data());
+        }
+
+      public:
+        void operator delete(void *p) { ::operator delete(p); }
+
+        static ManagedNuclBuffer *create(size_t nucls) {
+            void *mem = ::operator new(totalSizeToAlloc<ST>(Sequence::DataSize(nucls)));
+            return new (mem) ManagedNuclBuffer();
+        }
+
+        static ManagedNuclBuffer *create(size_t nucls, ST *data) {
+            void *mem = ::operator new(totalSizeToAlloc<ST>(Sequence::DataSize(nucls)));
+            return new (mem) ManagedNuclBuffer(nucls, data);
+        }
+
+        const ST *data() const { return getTrailingObjects<ST>(); }
+        ST *data() { return getTrailingObjects<ST>(); }
+    };
+
+    size_t size_ : 32;
+    size_t from_ : 31;
+    bool   rtl_  : 1;  // Right to left + complimentary (?)
+    llvm::IntrusiveRefCntPtr<ManagedNuclBuffer> data_;
+    std::shared_ptr<llvm::SparseBitVector<>> empty_nucls_ = nullptr;
```

Comment:
Ouch. Maybe include this into nuclbuffer then? So we won't have 2 counters in the single sequence? :)